### DISCUSSION TITLE
Linux compatibility and runtime

### DISCUSSION THREAD

**Discussion by medihack** [fo-dicom/fo-dicom#1482] [Q&A]
I am the maintainer of an open source DICOM web tool ([ADIT](https://github.com/radexperts/adit)). Under the hood, we currently use pynetdicom for transferring DICOM files, but we would like to replace it with something more performant. Probable candidates are DCMTK, Dcm4Che and last but not least Fellow Oak DICOM. I especially like the API of FO-DICOM, but I don't have any experience with C#. Are there any drawbacks when installing FO-DICOM in a linux container? It seems to me that I can use Mono as a runtime but also .NET core. What would you recommend? Besides the DICOM transfer itself, we would also like to stream images in the future (so we also need to support the codecs).

**Comment by mrbean-bremen** [comment]
While I like both `pynetdicom` and `fo-dicom`, if you want the best possible performance, you probably should go with `dcmtk`. That being said, maintenance efforts and ease of use are also a factor here, and in that respect I certainly would prefer `fo-dicom`. It also depends on your environment, of course. 

Some time ago I have done some performance tests with both `dcmtk` and `fo-dicom` regarding receiving data as a Store SCP. I don't remember the actual numbers, but `dcmtk` was about 1.5 times faster than `fo-dicom` with the given use case (mostly storing MR studies). This was with using an SSD and with both the sender and receiver on the same host, so the I/O overhead was mostly ignored here. Depending on your network and disk performance, this may change the relationship in favor of `fo-dicom`. Also, there have been recent performance and memory improvements in `fo-dicom` that have not been taken into account.
My conclusion at the time was that the performance difference is not large enough to compensate for the easier usage and maintenance, and we decided to use `fo-dicom`. 

I have to add that this was only tested under Windows with .NET Core - I have no experience with .NET Core/Mono under Linux. As far as I understand, .NET Core has no problems under Linux, at least as long as you don't use any GUI functionality. Running the server in a docker container is something that we also have considered as a future option. 

I can't say anything about the `Dcm4Che` performance, not having used it (I will not speculate about the probable performance of a Java application here...)

**Reply by medihack** [reply]
Thanks a lot for your insights. I will give it a try. In the [build notes](https://fo-dicom.github.io/stable/v5/developer/buildnotes.html) it says that tests _are targeting .Net Framework 4.6.2, .Net Core 2.1 and .Net Core 3.1. In order to run the unittests you need to have those SDKs installed_. 
But ` .Net Framework` seems to be something that is only available for Windows (or at least not for Linux). Does `fo-dicom` also contain some GUI stuff?

**Reply by mrbean-bremen** [reply]
I think that part of the documentation is indeed only meant for Windows, the wording may be a bit misleading. `fo-dicom` is built against NetStandard 2.0 and shall run fine under Linux and MacOS, but the CI tests are currently only executed under Windows against the mentioned targets. I think it could make sense to add also tests for Linux and maybe MacOS, this should not be too difficult.
And yes, there are some GUI parts in `fo-dicom` for displaying images (which you won't need). I haven't used them either, so I have no experience there.

**Comment by StevenBorg** [comment]
We use fo-DICOM in the Azure Health Data Services. Our code base is written in .NET and runs in Kubernetes orchestrated Linux containers. ---- On Sun, 11 Dec 2022 12:11:17 -0800  ***@***.******@***.***> wrote ---- 
I think that part is indeed only meant for Windows, the wording may be a bit misleading. fo-dicom is built against NetStandard 2.0, but the CI tests are currently only executed under Windows against the mentioned targets. I think it could make sense to add also tests for Linux and maybe MacOS, this should not be too difficult. 
And yes, there are some GUI parts in fo-dicom for displaying images (which you won't need). I haven't used them either, so I have no experience there.

**Comment by amoerie** [comment]
We have shoveled terabytes (might already be petabytes by now..) of DICOM over the internet using fo-dicom in the past few years. Even fo-dicom 4.x could be made fast enough with the right configuration. 
In my experience, the network and disk is usually more of a bottle neck than the CPU overhead of fo-dicom. 
And yes, I would expect the fo-dicom network layer to work completely under linux on .NET 6 or .NET 7. 
If you encounter issues, please let us know.

**Comment by jaime-olivares** [comment]
We are working already more than 5 years with fo-dicom over Ubuntu Server. Ingested 50+ TB in few days, no technical problems.
The performance can be increased by running multiple instances of your application.

**Comment by medihack** [comment]
Ultimately, we decided to continue using pynetdicom. Coming from a research background, we found it easier to find talent skilled in Python than in C#. Additionally, the ecosystem of Python is richer. Also as our app is very well parallelizable and does not demand real-time viewing of images, performance is not a significant concern. That said, fo-dicom seems to be a very good and stable project.
